tools/codesigningtool/codesigningtool.py

```diff
@@ -0,0 +1,105 @@
+# limitations under the License.
+
+import argparse
+import base64
+import json
+import os
+import plistlib
+import re
+import subprocess
+import sys
+
+
+def _check_output(args, inputstr=None):
+    proc = subprocess.Popen(args,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE)
+    return proc.communicate(input=inputstr)[0]
+
+
+def _parse_mobileprovision_file(mobileprovision_file):
+    """Reads and parses a mobileprovision file."""
+    plist_xml = subprocess.check_output([
+        "security", "cms",
+        "-D",
+        "-i", mobileprovision_file,
+    ])
+    if hasattr(plistlib, "readPlistFromString"):
+        # Python 2
+        return plistlib.readPlistFromString(plist_xml)
+    else:
+        # Python 3
+        return plistlib.readPlistFromBytes(plist_xml)
+
+
+def _certificate_fingerprint(identity):
+    """Extracts a fingerprint given identity bytes contained in a
+    mobileprovision file."""
+    fingerprint = _check_output([
+        "openssl", "x509", "-inform", "DER", "-noout", "-fingerprint",
+    ], inputstr=identity).strip().decode("utf-8")
+    fingerprint = fingerprint.replace("SHA1 Fingerprint=", "")
+    fingerprint = fingerprint.replace(":", "")
+    return fingerprint
+
+
+def _get_identities_from_provisioning_profile(mpf):
+    """Iterates through all the identities in a provisioning profile, lazily."""
+    for identity in mpf["DeveloperCertificates"]:
+        yield _certificate_fingerprint(identity.data)
+
+
+def _find_codesign_identities():
+    """Finds code signing identities on the current system."""
+    ids = []
+    output = _check_output([
+        "security", "find-identity", "-v", "-p", "codesigning",
+    ]).strip().decode("utf-8")
+    for line in output.splitlines():
+        m = re.search(r"([A-F0-9]{40})", line)
+        if m:
+            ids.append(m.group(0))
+    return ids
+
+
+def _find_codesign_identity(mobileprovision):
+    """Finds a valid codesign identity on the system given an mobileprovision
+    file."""
+    mpf = _parse_mobileprovision_file(mobileprovision)
+    ids_codesign = set(_find_codesign_identities())
+    for id_mpf in _get_identities_from_provisioning_profile(mpf):
+        if id_mpf in ids_codesign:
+            return id_mpf
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(description="codesign wrapper")
+    parser.add_argument("--mobileprovision", type=str, help="mobileprovision file")
+    parser.add_argument("--codesign", type=str, help="path to codesign binary")
+    parser.add_argument("--identity", type=str, help="specific identity to sign with")
+    args, codesign_args = parser.parse_known_args()
+    identity = args.identity
+    if identity is None:
+        identity = _find_codesign_identity(args.mobileprovision)
+    # No identity was found, fail
+    if identity == None:
+        print("Unable to find identity on the system matching the ones in %s" % args.mobileprovision)
```

Comment:
People don't always read logs, so we tend to start these with _ERROR: ..._, likewise, stick it on stderr to be more correct.